Improve visibility of setting adjustments on mod icons

peppy opened the pull request:
This adds two improvements:

## Show marker icon when any setting is non-default

![osu Game Tests 2025-04-28 at 09 09 31](https://github.com/user-attachments/assets/35b1f993-f87e-4304-a3ee-dbd72c02e57a)

![osu! 2025-04-28 at 09 11 36](https://github.com/user-attachments/assets/2a363eb5-062a-422c-ac7e-3baaf9c212dc)

## Show inline text when a single setting is adjusted via difficulty adjust mode

![osu Game Tests 2025-04-28 at 09 10 34](https://github.com/user-attachments/assets/6bff1229-804f-4107-a403-e4ecaf38856f)

![osu! 2025-04-28 at 09 12 43](https://github.com/user-attachments/assets/16b9d7b7-b54c-492d-bcf0-9bb2a2e7d69d)

---

peppy committed:
Adjust mod icon test scene to show overlapping versions too

peppy committed:
Show singular difficulty adjust modifications inline in mod icons

peppy committed:
Show marker when settings are adjusted

bdach commented on osu.Game/Rulesets/Mods/IMod.cs:
```diff
@@ -81,5 +83,29 @@ public interface IMod : IEquatable<IMod>
+        /// <summary>
+        /// Whether any user adjustable setting attached to this mod has a non-default value.
+        /// </summary>
+        bool HasNonDefaultSettings
```
I'd hope it's obvious to anyone that this is the *instantaneous* state and you need a setting tracker for this to work reliably in a context where mods are mutable, but maybe worth adding xmldoc to spell that out?

bdach commented on osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs:
```diff
@@ -67,6 +68,33 @@ protected ModDifficultyAdjust()
             }
         }
 
+        public virtual int AdjustedSettingsCount
```
Is there a reason this is virtual and not using the same reflection magic that `IMod.HasNonDefaultSettings` is? Is it for the catch "hard rock offsets" thingy or something?

peppy replied:
Hmm, I was just avoiding reflection here... but it turned out to be required for the other usage so maybe just bite the bullet and do the same here?

bdach replied:
I'd say so at this point. Not sure this is saving much.

sineplusx commented:
I dont know if my opinion is wanted here, but I think the marker icon is a bad design choice. At that small of a resolution it looks more like a sliced orange, and adds more clutter to the already stacked mod display. Maybe an outline could work better?

The icon seems unnecessary for rate adjust mods and difficulty adjust, where the modification is already apparent by the text beside it.

peppy commented:
Thanks for your feedback. You can open a discussion with a proposal (please include full visuals of how you'd see it looking).

I'm not open to further design feedback right now, and already sought internal feedback on this to the point I am happy with it.

> Maybe an outline could work better?

A border will definitely not be visible enough.

> The icon seems unnecessary for rate adjust mods and difficulty adjust, where the modification is already apparent by the text beside it.

Wasn't sure on this one but internal feedback said it was best to have it there. Can adjust based on user feedback after (at least) a week after release once everyone has had a chance to adjust their eyes and brains.

bdach commented on osu.Game/Rulesets/UI/ModIcon.cs:
```diff
@@ -207,6 +235,11 @@ private void updateMod(IMod value)
             backgroundColour = colours.ForModType(value.Type);
             updateColour();
 
+            if (mod.HasNonDefaultSettings)
+                adjustmentMarker.Show();
+            else
+                adjustmentMarker.Hide();
```
(This is the only real blocker) Case in previous point re: safety of reading this, this should be in the "update extended info" callback as it's hooked up to a change tracker:

https://github.com/ppy/osu/blob/3ee8eb2809b7570d315e823851214e47e9d64ba7/osu.Game/Rulesets/UI/ModIcon.cs#L189

peppy commented:
The first looks fine to my eye, as does the original. The position should not change though, and the cog icon is better than the star.

peppy committed:
Adjust colours to match icon better

peppy committed:
Add xmldoc note explaining new flag is instantaneous state

peppy committed:
Don't bother with localised implementation of `AdjustedSettingsCount`

I was avoiding using reflection to save on overheads, but it's probably
not worth it.

peppy committed:
Fix adjustment marker not updating when settings' states change

bdach commented on osu.Game/Rulesets/Mods/Mod.cs:
```diff
@@ -75,6 +75,27 @@ public abstract class Mod : IMod, IEquatable<Mod>, IDeepCloneable<Mod>
+        /// <summary>
+        /// The number of settings on this mod instance which have been adjusted by the user from their default values.
+        /// </summary>
+        public int UserAdjustedSettingsCount
```
Bit weird to have this on `Mod` given that the only usage (so far? ever?) is in `ModDifficultyAdjust`, and bit weird that `IMod.HasNonDefaultSettings` copy-pastes much of this, but I guess I don't have any real objections

bdach replied:
> I guess the count has less potential to be useful elsewhere so it should be preferred local?

I would go with this for now yeah. The bool flag can be global.

peppy committed:
Move `UserAdjustedSettingsCount` local to `ModDifficultyAdjust` in absence of other usages
